From mtosatti@redhat.com  Tue Feb 15 10:53:55 2011
From: Avi Kivity <avi@redhat.com>
Date: Fri, 14 Jan 2011 10:10:18 -0200
Subject: KVM: MMU: Fix incorrect direct gfn for unpaged mode shadow
To: greg@kroah.com, avi@redhat.com, mtosatti@redhat.com, stable@kernel.org
Message-ID: <20110114121048.294837491@amt.cnet>


From: Avi Kivity <avi@redhat.com>

commit c093b8b46c5f0dd12d799f0d6a3b579863df72f6 upstream.

We use the physical address instead of the base gfn for the four
PAE page directories we use in unpaged mode.  When the guest accesses
an address above 1GB that is backed by a large host page, a BUG_ON()
in kvm_mmu_set_gfn() triggers.

Resolves: https://bugzilla.kernel.org/show_bug.cgi?id=21962
Reported-and-tested-by: Nicolas Prochazka <prochazka.nicolas@gmail.com>
Signed-off-by: Avi Kivity <avi@redhat.com>
Cc: Marcelo Tosatti <mtosatti@redhat.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>
---
 arch/x86/kvm/mmu.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/arch/x86/kvm/mmu.c
+++ b/arch/x86/kvm/mmu.c
@@ -2271,7 +2271,7 @@ static int mmu_alloc_roots(struct kvm_vc
 			return 1;
 		if (tdp_enabled) {
 			direct = 1;
-			root_gfn = i << 30;
+			root_gfn = i << (30 - PAGE_SHIFT);
 		}
 		spin_lock(&vcpu->kvm->mmu_lock);
 		kvm_mmu_free_some_pages(vcpu);
